### ISSUE TITLE
Create Path.Windows and Path.Posix modules like in filepath

### ISSUE THREAD

**Issue by sjakobi** [commercialhaskell/path#82]
The module structure would look roughly like this:

* All the code is in Path.Internal and depends on two `isPosix` and `isWindows` constants for platform-specific stuff
* Path.Windows `#include`s Path.Internal, setting `isWindows` to `True` and `isPosix` to `False`
* Path.Posix `#include`s Path.Internal, setting `isWindows` to `False` and `isPosix` to `True`
* Path reexports Path.Windows or Path.Posix depending on platform

The main benefit would be that we could run platform-specific tests on _any_ platform and could easily check e.g. Windows behavior on Linux. As far as I can tell this this works nicely for [filepath](http://hackage.haskell.org/package/filepath-1.4.1.2).

**Comment by sjakobi** [comment]
Any comments or opinions before I give this a try?

**Comment by sjakobi** [comment]
I think this should also help us resolve #74 and #81.

**Comment by NorfairKing** [comment]
This sounds good to me!
You might want to use a datatype `data OS = Windows | Linux` to prevent messiness if we ever want to support more than these two OSes.
Could you do me a favor and check that GHC can compile-away any (constant) `if isWindows` expressions.

**Comment by sjakobi** [comment]
> You might want to use a datatype data OS = Windows | Linux to prevent messiness if we ever want to support more than these two OSes.

Should be `Windows | Posix` but I think I'll do it just like filepath does it – before we can support any OS that is neither Posix nor Windows, `filepath` will have to be fixed too.

> Could you do me a favor and check that GHC can compile-away any (constant) if isWindows expressions.

Seems to work:

```
~/s/filepath (master|…) $ stack ghc -- -O2 -ddump-simpl System/FilePath/Posix.hs  
[1 of 1] Compiling System.FilePath.Posix ( System/FilePath/Posix.hs, System/FilePath/Posix.o )

==================== Tidy Core ====================
Result size of Tidy Core
  = {terms: 1,705, types: 1,837, coercions: 24}

-- RHS size: {terms: 2, types: 0, coercions: 0}
pathSeparator :: Char
[GblId,
 Caf=NoCafRefs,
 Str=DmdType m,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 10 20}]
pathSeparator = GHC.Types.C# '/'#

-- RHS size: {terms: 2, types: 0, coercions: 0}
pathSeparators :: [Char]
[GblId,
 Str=DmdType,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=False, ConLike=False,
         WorkFree=False, Expandable=False, Guidance=IF_ARGS [] 40 0}]
pathSeparators = GHC.CString.unpackCString# "/"#
…
```

**Comment by NorfairKing** [comment]
> Should be Windows | Posix but I think I'll do it just like filepath does it – before we can support any OS that is neither Posix nor Windows, filepath will have to be fixed too.

Fair enough. Proceed :D

> >    Could you do me a favor and check that GHC can compile-away any (constant) if isWindows expressions.

> Seems to work:

Perfect, Thanks for checking!

**Comment by ygale** [comment]
The importance of this issue is far beyond just running test suites.

Currently, the path library is not a general-purpose replacement for filepath, because it does not support the common use case where your program is running on one platform and you need to process file path data that you received from another platform.

We need that all the time. And here is a recent [post on the haskell-beginners mailing list](https://mail.haskell.org/pipermail/beginners/2017-August/017761.html) on the same topic

This issue solves that problem.

@sjakobi are you still working on implementing this?

**Comment by mrkkrp** [comment]
Hey @ygale, I agree that this is an important issue and we should publish a new version with per-OS modules and automatic detection a la `filepath`.

If @sjakobi is busy I could find time to prepare a PR in timely manner.

**Comment by joehillen** [comment]
> Currently, the path library is not a general-purpose replacement for filepath, because it does not support the common use case where your program is running on one platform and you need to process file path data that you received from another platform.

I'm having trouble imagining how that would work. 

* Would there be different types for each platform, as in `data Path = WinPath | PosixPath`?
* How would serialization and deserialization work between the 2 platforms? What would be the intermediate representation?
* Is the problem that you can't parse and construct a Windows `Path` from a Posix `Path` and vice versa? How would that work with drive letters?
* What would prevent users from constructing a `Path` from the wrong platform and trying to write it to disk?

**Comment by sjakobi** [comment]
I'll try to get this done until next week, i.e. 2017-09-04.

@joehillen These are some good questions. I'd also expect users who want to manipulate e.g. Windows `Path`s on a POSIX platform to desire a few more amenities like newtype wrappers etc.

Regarding serialization: For starters it might be enough to export our `parseJSONWith` utility, in order to facilitate custom platform-specific `FromJSON` instances.

**Comment by sjakobi** [comment]
I've got most of the changes to the library done on [this branch](https://github.com/commercialhaskell/path/tree/issue82) but still need to adjust the docs. I'll try to wrap that part up tomorrow.

Regarding the changes to the testsuites I don't really have a design worked out. I think I'll address those in a second PR, probably after a bit of discussion.

**Comment by joehillen** [comment]
Resolved via #111
